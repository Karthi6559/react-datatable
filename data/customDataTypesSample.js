import React from "react";
import { TableCell } from "@material-ui/core";

const customDataTypesSample = [
  {
    dataType: "number",
    component: cellVal => (
      <TableCell>
        <div style={{ color: "orange" }}>{cellVal}</div>
      </TableCell>
    )
  },
  {
    dataType: "text",
    component: cellVal => (
      <TableCell>
        <div style={{ color: "purple" }}>{cellVal}</div>
      </TableCell>
    )
  },
  {
    dataType: "iban",
    component: cellVal => (
      <TableCell>
        <div style={{ color: "red" }}>{cellVal}</div>
      </TableCell>
    )
  }
];

export default customDataTypesSample;
