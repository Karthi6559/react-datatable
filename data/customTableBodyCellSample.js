import React from "react";
import { TableCell } from "@material-ui/core";
import { cellValPropType, columnPropType } from "../src/proptypes";

const customTableBodyCellSample = ({ cellVal, column }) => {
  let val;
  switch (column.dataType) {
    case "boolean":
      if (cellVal) {
        val = (
          <div
            className="data-boolean"
            style={{ color: "green", textAlign: "center" }}
          >
            Yes
          </div>
        );
      } else {
        val = <div style={{ color: "red", textAlign: "center" }}>No</div>;
      }
      break;
    default:
      val = (
        <div className="data" style={{ color: "blue" }}>
          {cellVal}
        </div>
      );
      break;
  }
  if (val) {
    return <TableCell>{val}</TableCell>;
  }
  return <TableCell className="no-data" />;
};

customTableBodyCellSample.propTypes = {
  cellVal: cellValPropType,
  column: columnPropType
};

export default customTableBodyCellSample;
