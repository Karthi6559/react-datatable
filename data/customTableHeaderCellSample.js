import React from "react";
import { columnPropType } from "../src/proptypes";

const customTableHeaderCellSample = ({ column }) => {
  switch (column.dataType) {
    case "number":
      return <div className="number">{column.label}</div>;
    case "text":
      return <div className="text">{column.label}</div>;
    case "boolean":
      return <div className="boolean">{column.label}</div>;
    case "date":
      return <div className="date">{column.label}</div>;
    default:
      return <div className="default">{column.label}</div>;
  }
};

customTableHeaderCellSample.propTypes = {
  column: columnPropType
};

export default customTableHeaderCellSample;
