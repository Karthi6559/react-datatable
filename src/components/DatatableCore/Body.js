import React, { Component } from "react";
import { connect } from "react-redux";
import { TableBody, TableRow, TableCell } from "@material-ui/core";
import { rowsPropType } from "../proptypes/proptypes";

class Body extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  render() {
    const { rows } = this.props;
    return (
      <TableBody>
        {rows.map(row => (
          <TableRow key={row.id}>
            {Object.keys(row).map((key, id) => (
              <TableCell
                key={key}
                align={id > 0 ? "right" : "left"}
                component={id === 0 ? "th" : undefined}
                scope={id === 0 ? "row" : undefined}
              >
                <div>{row[key]}</div>
              </TableCell>
            ))}
          </TableRow>
        ))}
      </TableBody>
    );
  }
}

Body.propTypes = {
  rows: rowsPropType
};

const mapStateToProps = state => {
  return {
    rows: state.datatableReducer.data.rows
  };
};

export default connect(mapStateToProps)(Body);
