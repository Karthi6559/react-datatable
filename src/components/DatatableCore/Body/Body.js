import React, { Component, Fragment } from "react";
import { findDOMNode } from "react-dom";
import { connect } from "react-redux";
import { ScrollSyncPane } from "react-scroll-sync";
import { FixedSizeList } from "react-window";
import { throttle } from "lodash";
import BodyRow from "./BodyRow";
import { setIsScrolling as setIsScrollingAction } from "../../../redux/actions/datatableActions";
import {
  rowsPropType,
  CustomTableBodyRowPropType,
  columnsOrderPropType,
  dimensionsPropType,
  setIsScrollingPropType,
  isScrollingPropType,
  keyColumnPropType,
  rowsEditedPropType,
  heightNumberPropType,
  widthNumberPropType,
  columnSizeMultiplierPropType
} from "../../../proptypes";

export const tableRef = React.createRef();

export class Body extends Component {
  componentDidMount() {
    const virtualizedContainer = findDOMNode(tableRef.current);
    const callBack = () =>
      throttle(() => this.handleScroll(virtualizedContainer.scrollLeft), 500);
    virtualizedContainer.addEventListener("scroll", callBack());
  }

  componentWillUnmount() {
    findDOMNode(tableRef.current).removeEventListener("scroll", null);
  }

  handleScroll = scrollLeft => {
    const { setIsScrolling, isScrolling } = this.props;
    const bool = scrollLeft > 0;
    if (bool !== isScrolling) {
      setIsScrolling(bool);
    }
  };

  rowBuilder = ({ index, style }) => {
    const {
      CustomTableBodyRow,
      rows,
      dimensions,
      columnsOrder,
      keyColumn,
      rowsEdited
    } = this.props;
    const key = `row-${index}`;
    const { columnSizeMultiplier } = dimensions;
    let row = rows[index];
    row = rowsEdited.find(r => r[keyColumn] === row[keyColumn]) || row;
    const editing =
      rowsEdited.find(r => r[keyColumn] === row[keyColumn]) !== undefined;

    if (CustomTableBodyRow !== null) {
      return (
        <div
          style={{
            top: style.top,
            height: style.height,
            position: style.position,
            borderBottom: "1px solid rgba(224, 224, 244, 1)"
          }}
        >
          <CustomTableBodyRow
            row={row}
            columnsOrder={columnsOrder}
            rowIndex={index}
            height={style.height}
            columnSizeMultiplier={columnSizeMultiplier}
            key={key}
          />
        </div>
      );
    }

    return <BodyRow row={row} editing={editing} style={style} key={key} />;
  };

  render() {
    const {
      rows,
      dimensions,
      columnsOrder,
      rowsEdited,
      height,
      width,
      totalWidthNumber,
      columnSizeMultiplier
    } = this.props;

    return (
      <Fragment>
        {rows.length > 0 && (
          <div className="Table-Body">
            <ScrollSyncPane>
              <FixedSizeList
                ref={tableRef}
                className="virtualized-container"
                height={dimensions.body.heightNumber}
                itemCount={rows.length}
                itemSize={dimensions.row.heightNumber}
                width={dimensions.datatable.widthNumber}
                columnsOrder={columnsOrder}
                rowsEdited={rowsEdited}
                rows={rows}
              >
                {this.rowBuilder}
              </FixedSizeList>
            </ScrollSyncPane>
          </div>
        )}

        {rows.length === 0 && (
          <Fragment>
            <div
              id="no-rows"
              style={{ height: height - 15, width: width - 15 }}
            >
              There is no result for your search
            </div>
            <ScrollSyncPane>
              <div
                style={{
                  overflowX: columnSizeMultiplier === 1 ? "scroll" : "hidden",
                  overflowY: "hidden",
                  height: "15px",
                  width: width - 15
                }}
              >
                <div
                  style={{
                    width: totalWidthNumber
                  }}
                >
                  .
                </div>
              </div>
            </ScrollSyncPane>
          </Fragment>
        )}
      </Fragment>
    );
  }
}

Body.propTypes = {
  rows: rowsPropType.isRequired,
  columnsOrder: columnsOrderPropType.isRequired,
  dimensions: dimensionsPropType.isRequired,
  CustomTableBodyRow: CustomTableBodyRowPropType,
  setIsScrolling: setIsScrollingPropType,
  isScrolling: isScrollingPropType.isRequired,
  keyColumn: keyColumnPropType.isRequired,
  rowsEdited: rowsEditedPropType.isRequired,
  height: heightNumberPropType.isRequired,
  width: widthNumberPropType.isRequired,
  totalWidthNumber: widthNumberPropType,
  columnSizeMultiplier: columnSizeMultiplierPropType
};

const mapDispatchToProps = dispatch => {
  return {
    setIsScrolling: bool => dispatch(setIsScrollingAction(bool))
  };
};

const mapStateToProps = state => {
  return {
    rows: state.datatableReducer.pagination.rowsCurrentPage,
    dimensions: state.datatableReducer.dimensions,
    columnsOrder:
      state.datatableReducer.features.userConfiguration.columnsOrder,
    rowsEdited: state.datatableReducer.rowsEdited,
    height: state.datatableReducer.dimensions.body.heightNumber,
    width: state.datatableReducer.dimensions.datatable.widthNumber,
    keyColumn: state.datatableReducer.keyColumn,
    isScrolling: state.datatableReducer.dimensions.isScrolling,
    totalWidthNumber:
      state.datatableReducer.dimensions.datatable.totalWidthNumber,
    columnSizeMultiplier:
      state.datatableReducer.dimensions.columnSizeMultiplier,
    CustomTableBodyRow: state.customComponentsReducer.CustomTableBodyRow
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Body);
