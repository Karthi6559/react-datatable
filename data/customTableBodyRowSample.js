import React from "react";
import {
  rowPropType,
  columnsOrderPropType,
  indexPropType,
  heightNumberPropType,
  columnSizeMultiplierPropType
} from "../src/proptypes";
import { simpleOptionsSample } from "./samples";

const customTableBodyRowSample = ({
  row,
  columnsOrder,
  rowIndex,
  columnSizeMultiplier,
  height
}) => {
  const { columns } = simpleOptionsSample.data;
  const columnAction = {
    id: "actions",
    label: "Actions",
    colSize: "150px",
    editable: false
  };
  return (
    <div
      className={
        rowIndex % 2 === 0 ? "Table-Row stripped" : "Table-Row not-stripped"
      }
      style={{
        height,
        background: rowIndex % 2 === 0 ? "#3f3f3f" : "#5f5f5f"
      }}
    >
      {columnsOrder.map(columnId => {
        const column =
          columnId === "actions"
            ? columnAction
            : columns.find(col => col.id === columnId);
        const width = `${(
          column.colSize.split("px")[0] * columnSizeMultiplier
        ).toString()}px`;
        return (
          <div className="Table-Cell" key={columnId}>
            <div style={{ color: "pink", width }}>{row[columnId]}</div>
          </div>
        );
      })}
    </div>
  );
};

customTableBodyRowSample.propTypes = {
  row: rowPropType,
  columnsOrder: columnsOrderPropType,
  rowIndex: indexPropType,
  height: heightNumberPropType,
  columnSizeMultiplier: columnSizeMultiplierPropType
};

export default customTableBodyRowSample;
