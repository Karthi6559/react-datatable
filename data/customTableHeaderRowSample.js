import React from "react";
import {
  columnsOrderPropType,
  columnSizeMultiplierPropType
} from "../src/proptypes";
import { simpleOptionsSample } from "./samples";

const customTableHeaderRowSample = ({ columnsOrder, columnSizeMultiplier }) => {
  const { columns } = simpleOptionsSample.data;
  const columnAction = {
    id: "o2xpActions",
    label: "Actions",
    colSize: "150px",
    editable: false
  };
  return (
    <div className="Table-Row">
      {columnsOrder.map(columnId => {
        const column =
          columnId === "o2xpActions"
            ? columnAction
            : columns.find(col => col.id === columnId);
        const width = `${(
          column.colSize.split("px")[0] * columnSizeMultiplier
        ).toString()}px`;
        return (
          <div className="Table-Header-Cell" key={column.id}>
            <div style={{ color: "green", width }}>{column.label}</div>
          </div>
        );
      })}
    </div>
  );
};

customTableHeaderRowSample.propTypes = {
  columnsOrder: columnsOrderPropType,
  columnSizeMultiplier: columnSizeMultiplierPropType
};

export default customTableHeaderRowSample;
