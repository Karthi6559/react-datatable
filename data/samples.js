import storeSample from "./storeSample";
import storeNoCustomComponentsSample from "./storeNoCustomComponentsSample";
import storeCustomTableBodyCellComponentSample from "./storeCustomTableBodyCellComponentSample";
import storeCustomTableBodyRowComponentSample from "./storeCustomTableBodyRowComponentSample";
import storeCustomTableHeaderCellComponentSample from "./storeCustomTableHeaderCellComponentSample";
import storeCustomTableHeaderRowComponentSample from "./storeCustomTableHeaderRowComponentSample";
import storeNoDataSample from "./storeNoDataSample";
import storeNoRowsDataSample from "./storeNoRowsDataSample";
import simpleOptionsSample from "./simpleOptionsSample";
import simpleOptionsNoDataSample from "./simpleOptionsNoDataSample";
import maximumOptionsSample from "./maximumOptionsSample";
import minimumOptionsSample from "./minimumOptionsSample";
import defaultOptionsSample from "./defaultOptionsSample";
import mergedSimpleOptionsSample from "./mergedSimpleOptionsSample";
import mergedMaximumOptionsSample from "./mergedMaximumOptionsSample";
import mergedMinimumOptionsSample from "./mergedMinimumOptionsSample";
import customTableBodyRowSample from "./customTableBodyRowSample";
import customTableBodyCellSample from "./customTableBodyCellSample";
import customTableHeaderRowSample from "./customTableHeaderRowSample";
import customTableHeaderCellSample from "./customTableHeaderCellSample";
import customDataTypesSample from "./customDataTypesSample";

export {
  storeSample,
  storeNoCustomComponentsSample,
  storeCustomTableBodyCellComponentSample,
  storeCustomTableBodyRowComponentSample,
  storeCustomTableHeaderCellComponentSample,
  storeCustomTableHeaderRowComponentSample,
  storeNoDataSample,
  storeNoRowsDataSample,
  simpleOptionsSample,
  simpleOptionsNoDataSample,
  maximumOptionsSample,
  minimumOptionsSample,
  defaultOptionsSample,
  mergedSimpleOptionsSample,
  mergedMaximumOptionsSample,
  mergedMinimumOptionsSample,
  customTableBodyRowSample,
  customTableBodyCellSample,
  customTableHeaderRowSample,
  customTableHeaderCellSample,
  customDataTypesSample
};
