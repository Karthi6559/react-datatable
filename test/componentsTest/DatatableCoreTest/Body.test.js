import React from "react";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import Body from "../../../src/components/DatatableCore/Body";
import { storeSample } from "../../data/samples";

describe("Body component", () => {
  const mockStore = configureStore();
  beforeEach(() => {});
  it("should render without errors", () => {
    const store = mockStore(storeSample);
    const tree = renderer
      .create(
        <Provider store={store}>
          <Body />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
