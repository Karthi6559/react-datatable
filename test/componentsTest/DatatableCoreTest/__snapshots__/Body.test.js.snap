// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Body component should render without errors 1`] = `
<tbody
  className="MuiTableBody-root-1"
>
  <tr
    className="MuiTableRow-root-2"
  >
    <th
      aria-sort={null}
      className="MuiTableCell-root-7 MuiTableCell-body-9 MuiTableCell-alignLeft-15"
      scope="row"
    >
      <div>
        50cf
      </div>
    </th>
    <td
      aria-sort={null}
      className="MuiTableCell-root-7 MuiTableCell-body-9 MuiTableCell-alignRight-17"
    >
      <div>
        28
      </div>
    </td>
    <td
      aria-sort={null}
      className="MuiTableCell-root-7 MuiTableCell-body-9 MuiTableCell-alignRight-17"
    >
      <div>
        Kerr Mayo
      </div>
    </td>
    <td
      aria-sort={null}
      className="MuiTableCell-root-7 MuiTableCell-body-9 MuiTableCell-alignRight-17"
    >
      <div />
    </td>
    <td
      aria-sort={null}
      className="MuiTableCell-root-7 MuiTableCell-body-9 MuiTableCell-alignRight-17"
    >
      <div>
        1972-09-04T11:09:59
      </div>
    </td>
  </tr>
  <tr
    className="MuiTableRow-root-2"
  >
    <th
      aria-sort={null}
      className="MuiTableCell-root-7 MuiTableCell-body-9 MuiTableCell-alignLeft-15"
      scope="row"
    >
      <div>
        209
      </div>
    </th>
    <td
      aria-sort={null}
      className="MuiTableCell-root-7 MuiTableCell-body-9 MuiTableCell-alignRight-17"
    >
      <div>
        34
      </div>
    </td>
    <td
      aria-sort={null}
      className="MuiTableCell-root-7 MuiTableCell-body-9 MuiTableCell-alignRight-17"
    >
      <div>
        Freda Bowman
      </div>
    </td>
    <td
      aria-sort={null}
      className="MuiTableCell-root-7 MuiTableCell-body-9 MuiTableCell-alignRight-17"
    >
      <div />
    </td>
    <td
      aria-sort={null}
      className="MuiTableCell-root-7 MuiTableCell-body-9 MuiTableCell-alignRight-17"
    >
      <div>
        1988-03-14T09:03:19
      </div>
    </td>
  </tr>
  <tr
    className="MuiTableRow-root-2"
  >
    <th
      aria-sort={null}
      className="MuiTableCell-root-7 MuiTableCell-body-9 MuiTableCell-alignLeft-15"
      scope="row"
    >
      <div>
        2dd81ef
      </div>
    </th>
    <td
      aria-sort={null}
      className="MuiTableCell-root-7 MuiTableCell-body-9 MuiTableCell-alignRight-17"
    >
      <div>
        14
      </div>
    </td>
    <td
      aria-sort={null}
      className="MuiTableCell-root-7 MuiTableCell-body-9 MuiTableCell-alignRight-17"
    >
      <div>
        Becky Lawrence
      </div>
    </td>
    <td
      aria-sort={null}
      className="MuiTableCell-root-7 MuiTableCell-body-9 MuiTableCell-alignRight-17"
    >
      <div />
    </td>
    <td
      aria-sort={null}
      className="MuiTableCell-root-7 MuiTableCell-body-9 MuiTableCell-alignRight-17"
    >
      <div>
        1969-02-10T04:02:44
      </div>
    </td>
  </tr>
</tbody>
`;
